### PR TITLE
Poc/more series with nullpoints

### PR DESCRIPTION
**Pull request by MarkusBarstad**
Added support [nullInteraction](https://api.highcharts.com/highcharts/plotOptions.series.nullInteraction) for more series types. Null points will now support data labels and tooltips.

Note: might be to early to request this being merged, but I feel like it would be good to have some other eyes looking at this.

Some things to note which could be better is:
- FocusBorder for null-points does not quite look right and for columns it should maybe have dimensions which reflect those of existing columns
- No default tooltip formatting for null points (in other words you need to supply a null-format). Should we have this?
- Should we have unit-tests for keyboard navigation here?

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1208158988997384

### CODE REVIEW

**Review by KacperMadej — COMMENTED** [review]
- tests needs demo.details (to run in the utils) and demo.css (at least for consistency)
- extend TypeScript definitions instead of casting to any for accurate type support, even in prototyping. For quick result-focused tests, consider a JSFiddle PoC using JS.

---

> FocusBorder for null-points does not quite look right and for columns it should maybe have dimensions which reflect those of existing columns

Looks like an a11y question. IMHO I would expect no focus box when there's nothing to focus on.

> No default tooltip formatting for null points (in other words you need to supply a null-format). Should we have this?

No need if explained in the docs next to the `nullInteraction` when that option is `false` by default.

> Should we have unit-tests for keyboard navigation here?

Looks like an a11y question, but I would say add null point testing in the keyboard navigation tests (or another more relevant a11y category).

---
General thoughts:
- internal a11y consultation is needed.
- more tests or demos to play with more series types would be nice to have - they will be needed for the testing in the end anyway.
- maybe this doesn't have to be perfect in every way from the very beginning? If it will work in most cases, tests, and demos then it could be improved on user request - based on the feedback.

**Comment by MarkusBarstad** [comment]
Alright, so now:

- Casts are gone, using declare module instead (unsure if this is done the best way)
- Unit-test expanded to multiple series (also added demo.details and demo.css)
- Focus border matches that of normal points (requested by a11y team)
- Simple demo for the option
- Docs for the option (unsure if this is properly done, cannot run doc-server locally)


Unsure if there should be more demos, but for testing I do have this fiddle for trying multiple seriestypes + tooltip-settings: https://jsfiddle.net/mkb93/8ust1kL3/

File: ts/Series/Line/LineSeries.ts [hunk 1]
```diff
@@ -374,6 +378,20 @@ export default LineSeries;
...
+ * When this is set to `true`, tooltips may highlight these points, and this
+ * option also enables keyboard navigation for such points. Formatting options
+ * for such points include [`nullFormat`](#tooltip.nullFormat) and [`nullFormater`](#tooltip.nullFormatter)
+ *
+ * @sample {highcharts} highcharts/series/null-interaction/
+ *         Chart with interactive `null` points
+ *
+ * @extends   plotOptions.series
+ * @product   highcharts highstock
+ * @apioption plotOptions.nullInteraction
```
[the remark this answers is not in the document]
Comment (hubertkozik) [reply ts/Series/Line/LineSeries.ts:392]: Also, this doclet is now declared in `LineSeries.ts`. I believe this feature works for other series, right?

File: ts/Accessibility/Components/SeriesComponent/SeriesDescriber.ts [hunk 2]
```diff
@@ -63,6 +63,12 @@ const {
  *
  * */
 
+declare module '../../../Core/Series/SeriesOptions' {
+    interface SeriesOptions {
+        nullInteraction?: boolean;
+    }
+}
+
```
Comment (hubertkozik) [inline ts/Accessibility/Components/SeriesComponent/SeriesDescriber.ts:71]: `nullInteraction` is declared in multiple files - here and in `ts/Accessibility/Components/SeriesComponent/SeriesKeyboardNavigation.ts`, `ts/Core/Pointer.ts`, `ts/Core/Series/DataLabel.ts`, `ts/Core/Series/Series.ts` is it necessary? Can we avoid it and declare it in one place? I think it should be declared once for Series.

File: ts/Extensions/Boost/BoostSeries.ts [hunk 3]
```diff
@@ -1259,7 +1261,7 @@ function seriesRenderCanvas(this: Series): void {
                 y = (d as any)[1];
             } else {
                 x = d as any;
-                y = yData[i] as any;
+                y = yData[i] as any ?? nullYSubstitute ?? null;
```
Comment (hubertkozik) [inline ts/Extensions/Boost/BoostSeries.ts:1264]: Cast as `any` is so close, worth removing 😁

File: ts/Series/Line/LineSeries.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (MarkusBarstad) [reply ts/Series/Line/LineSeries.ts:392]: Ah wow, should not have included @extends! 

I also changed @apioption to `plotOptions.line.nullInteraction`, is this correct? 🤔 

The option is intended for line series and its inheritors which is why I put it here 🤔 

Demo pages are not working for me so I cannot test this stuff unfortunately :grimacing:

File: ts/Series/Line/LineSeries.ts [hunk 4]
```diff
@@ -384,7 +402,7 @@ export default LineSeries;
  *         Line chart
  *
  * @extends   plotOptions.series
- * @product   highcharts highstock
+ * @product   highcharts
```
[the remark this answers is not in the document]
Comment (MarkusBarstad) [reply ts/Series/Line/LineSeries.ts:405]: My bad lol :grimacing: I should have read the diff more closely when pushing this

File: ts/Accessibility/Components/SeriesComponent/SeriesDescriber.ts (hunk 2 above)
Comment (MarkusBarstad) [reply ts/Accessibility/Components/SeriesComponent/SeriesDescriber.ts:71]: I am not entirely sure about this part :grimacing:

Option is inteded for LineSeries and series which extend it. Any ideas on how best to declare it? 🤔

File: ts/Series/Timeline/TimelineSeries.ts [hunk 5]
```diff
@@ -239,12 +239,16 @@ class TimelineSeries extends LineSeries {
 
     public getVisibilityMap(): Array<(boolean|TimelinePoint|TimelinePointOptions)> {
         const series = this,
+            nullInteraction = series.options.nullInteraction,
```
Comment (KacperMadej) [inline ts/Series/Timeline/TimelineSeries.ts:242]: Could you add a demo for the API and a test for the timeline series? When using just `null` it fails silently - it works when used like `{y: null}` but that is not very intuicial for a timeline series IMHO.
image

**Review by KacperMadej — COMMENTED** [review]
API docs in `nullFormatter` and `nullFormat` have "Does not work with series that don't display null points, like line, column, bar or pie." and if this will no longer be the case it should be updated. Not sure if there are any other parts of API or general docs that should also be updated - worth checkin.

File: ts/Series/Timeline/TimelineSeries.ts (hunk 5 above)
Comment (KacperMadej) [reply ts/Series/Timeline/TimelineSeries.ts:243]: This is still the case - there's no demo for the timeline series using null points linked in the API docs. It either looks wrong or doesn't seem to be working correctly (tooltip not showing, unable to hover) depending on using `null` or `{y:null}` as a point.

**Review by KacperMadej — CHANGES_REQUESTED** [review]

File: ts/Series/Timeline/TimelinePoint.ts~
Comment (hubertkozik) [inline ts/Series/Timeline/TimelinePoint.ts~:1]: This file shouldn't be added, right?

File: ts/Series/Timeline/TimelineSeries.ts~
Comment (hubertkozik) [inline ts/Series/Timeline/TimelineSeries.ts~:1]: Also, should not be added.

File: ts/Core/Series/SeriesDefaults.ts [hunk 6]
```diff
@@ -552,6 +552,20 @@ const seriesDefaults: PlotOptionsOf<Series> = {
...
+     * When this is set to `true`, tooltips may highlight these points, and this
+     * option also enables keyboard navigation for such points. Format options
+     * for such points include [`nullFormat`](#tooltip.nullFormat) and [`nullFormater`](#tooltip.nullFormatter)
+     *
+     * @sample {highcharts} highcharts/series/null-interaction/
+     *         Chart with interactive `null` points
+     *
+     * @type      {boolean|undefined}
+     * @product   highcharts highstock
+     * @apioption plotOptions.line.nullInteraction
```
Comment (hubertkozik) [inline ts/Core/Series/SeriesDefaults.ts:566]: This should be `@apioption plotOptions.series.nullInteraction`, because we can use it for all series? And should be excluded for series like pie.

**Review by hubertkozik — CHANGES_REQUESTED** [review]
Few new comments from me, looking better and better!

**Comment by MarkusBarstad** [comment]
@hubertkozik 
Regarding excluding certain series types from the option in the docs,
other options in `SeriesDefaults.ts` seem to just state what series are supported (like the `marker` option)

Therefore note that I have just stated this as well in the JSDoc field for `plotOptions.series.nullIntercaction`

**Comment by MarkusBarstad** [comment]
@KacperMadej @hubertkozik , Ok i think i healed this branch lol

All tests running as they should, and the branch does not delete any files.

**Review by KacperMadej — APPROVED** [review]

File: ts/Series/Column/ColumnSeries.ts [hunk 7]
```diff
@@ -636,9 +643,9 @@ class ColumnSeries extends Series {
                 // #3169, drilldown from null must have a position to work from.
                 // #6585, dataLabel should be placed on xAxis, not floating in
                 // the middle of the chart.
-                point.isNull ? translatedThreshold : barY,
-                barW,
-                point.isNull ? 0 : barH
+                point.isNull ? translatedThreshold - nullMinSpace : barY,
+                barW || point.isNull && nullMinSpace || 0,
```
Comment (TorsteinHonsi) [inline ts/Series/Column/ColumnSeries.ts:647]: Seems like `barW` is already correct for null points?

File: ts/Series/Column/ColumnSeries.ts [hunk 8]
```diff
@@ -494,6 +494,13 @@ class ColumnSeries extends Series {
         const series = this,
             chart = series.chart,
             options = series.options,
+            // For points whithout graphics (null points) this value is used
+            // to reserve space around the point such that:
+            //      - normal/null points are spaced similarily,
+            //      - focusborders of null points are like those of "0" points
+            // This ensures consistent dimensions between null/normal points.
+            minimumSpace = 8,
+            nullMinSpace = options.nullInteraction && minimumSpace || 0,
```
Comment (TorsteinHonsi) [inline ts/Series/Column/ColumnSeries.ts:503]: No need for a stepping stone?
```suggestion
            nullMinSpace = options.nullInteraction ? 8 : 0,
```

File: ts/Series/Column/ColumnSeries.ts [hunk 9]
```diff
@@ -636,9 +643,9 @@ class ColumnSeries extends Series {
                 // #3169, drilldown from null must have a position to work from.
                 // #6585, dataLabel should be placed on xAxis, not floating in
                 // the middle of the chart.
-                point.isNull ? translatedThreshold : barY,
-                barW,
-                point.isNull ? 0 : barH
+                point.isNull ? translatedThreshold - nullMinSpace : barY,
```
Comment (TorsteinHonsi) [inline ts/Series/Column/ColumnSeries.ts:646]: The `barY` and `barH` variables are already corrected for `minPointLength`, which is the exact equivalent of `nullMinSpace` but for zero points. See if we can combine those, at least do the corrections in the same place.

File: ts/Series/Column/ColumnSeries.ts [hunk 10]
```diff
@@ -690,7 +697,9 @@ class ColumnSeries extends Series {
             strokeWidth = (point && (point as any)[strokeWidthOption]) ||
                 (options as any)[strokeWidthOption] ||
                 (this as any)[strokeWidthOption] || 0,
-            opacity = pick(point && point.opacity, options.opacity, 1);
+            opacity = (point?.isNull && this.options.nullInteraction) ?
```
Comment (TorsteinHonsi) [inline ts/Series/Column/ColumnSeries.ts:700]:
```suggestion
            opacity = (point?.isNull && options.nullInteraction) ?
```

File: ts/Series/Column/ColumnSeries.ts [hunk 11]
```diff
@@ -690,7 +697,9 @@ class ColumnSeries extends Series {
...
+                0 :
+                pick(point && point.opacity, options.opacity, 1);
```
Comment (TorsteinHonsi) [inline ts/Series/Column/ColumnSeries.ts:702]:
```suggestion
                (point?.opacity ?? options.opacity ?? 1);
```

**Review by TorsteinHonsi — CHANGES_REQUESTED** [review]
Thanks, great work!

Extra thorough PR review from me, since this very much affects the core de la core. Will probably nitpick more in next round too.

File: ts/Core/Series/DataLabel.ts [hunk 12]
```diff
@@ -564,6 +564,11 @@ namespace DataLabel {
         fireEvent(this, 'drawDataLabels');
 
         if (series.hasDataLabels?.()) {
+            const nullLabel = (
+                seriesOptions?.nullInteraction &&
+                'Null'
+            );
```
Comment (TorsteinHonsi) [inline ts/Core/Series/DataLabel.ts:570]: I see this part was removed, then 8 diffs appeared, and a new commit named "Accommodating map series" re-introduced it. What happened? When I remove the `nullLabel` logic, I have no visual diffs in the maps section locally. Also, if the problem applies only to the map series, we should address it in the map code, not the core, for example by setting a default `dataLabels.nullFormatter`.

File: ts/Core/Series/Point.ts [hunk 13]
```diff
@@ -710,6 +710,10 @@ class Point {
 
         this.resolveColor();
 
+        if (typeof this.dataLabelOnNull === 'undefined') {
+            this.dataLabelOnNull = series.options.nullInteraction;
+        }
```
Comment (TorsteinHonsi) [inline ts/Core/Series/Point.ts:715]: Terser:
```suggestion
        this.dataLabelOnNull ??= series.options.nullInteraction;
```

**Review by TorsteinHonsi — CHANGES_REQUESTED** [review]

File: ts/Core/Series/DataLabel.ts (hunk 12 above)
Comment (MarkusBarstad) [reply ts/Core/Series/DataLabel.ts:570]: Oh sorry - this was supposed to be gone from the PR as was the logic using it.

**Review by hubertkozik — APPROVED** [review]
LGTM. Great work!

**Review by TorsteinHonsi — CHANGES_REQUESTED** [review]

File: samples/highcharts/series/null-interaction/demo.html [hunk 14]
```diff
@@ -0,0 +1,5 @@
+<script src="https://code.highcharts.com/highcharts.js"></script>
+<script src="https://code.highcharts.com/modules/accessiblity.js"></script>
```
Comment (marvin19) [inline samples/highcharts/series/null-interaction/demo.html:2]: Small typo, `accessiblity` should be `accessibility`

**Review by marvin19 — COMMENTED** [review]
My main concern is that we have the option [keyboardNavigation.seriesNavigation.skipNullPoints](https://api.highcharts.com/highcharts/accessibility.keyboardNavigation.seriesNavigation), which defaults to `true`, but it doesn’t seem to be working as expected as null points are not ignored for keyboard nav. Perhaps the unit test for keyboard navigation could be updated to check this behaviour, and a fix could be applied accordingly? Thoughts @TorsteinHonsi ?

**Comment by TorsteinHonsi** [comment]
@marvin19 Oh. You mean that in the combination of `nullInteraction` and `skipNullPoints` is true, `skipNullPoints` takes precedence? One solution could be to change the default of `skipNullPoints` to `undefined`, so that unless explicitly set, it will be the opposite of `nullInteraction`. Maybe do a quick discussion in the core team daily?

**Review by TorsteinHonsi — CHANGES_REQUESTED** [review]
After internal discussion we agreed that:
* `skipNullPoints` should be changed to `undefined` by default
* When `nullInteraction` is true, `skipNullPoints` (if undefined) behaves as if it were false
* When `nullInteraction` is false, `skipNullPoints` (if undefined) behaves as if it were true

File: ts/Accessibility/Components/SeriesComponent/SeriesKeyboardNavigation.ts [hunk 15]
```diff
@@ -134,18 +134,28 @@ function isSkipSeries(
...
+        skipNullPoints = a11yOptions
+            .keyboardNavigation
+            .seriesNavigation
+            .skipNullPoints;
+    /*
+    If (point.isNull) {
+        if (defined(skipNullPoints)) {
+            return skipNullPoints;
+        }
+    }*/
```
Comment (TorsteinHonsi) [inline ts/Accessibility/Components/SeriesComponent/SeriesKeyboardNavigation.ts:152]: This comment should probably be removed?

File: ts/Accessibility/Options/A11yDefaults.ts [hunk 16]
```diff
@@ -706,7 +706,7 @@ const Options: DeepPartial<A11yOptions> = {
                  *
                  * @since 8.0.0
                  */
-                skipNullPoints: true,
+                skipNullPoints: void 0,
```
Comment (TorsteinHonsi) [inline ts/Accessibility/Options/A11yDefaults.ts:709]: The description currently says "Skip null points when navigating through points with the keyboard". But we should also mention that the default is the opposite of the `nullInteraction` setting.

**Review by TorsteinHonsi — CHANGES_REQUESTED** [review]

**Review by TorsteinHonsi — APPROVED** [review]
